src/ILCompiler.Compiler/src/Compiler/PreInitFieldInfo.cs

```diff
@@ -8,9 +8,70 @@
 
 using Internal.TypeSystem;
 using Internal.TypeSystem.Ecma;
+using ILCompiler.DependencyAnalysis;
 
 namespace ILCompiler
 {
+    public abstract class PreInitFixupInfo : IComparable<PreInitFixupInfo>
+    {
+        /// <summary>
+        /// Offset into the blob
+        /// </summary>
+        public int Offset { get; }
+
+        public PreInitFixupInfo(int offset)
+        {
+            Offset = offset;
+        }
+
+        int IComparable<PreInitFixupInfo>.CompareTo(PreInitFixupInfo other)
+        {
+            return this.Offset - other.Offset;
+        }
+
+        /// <summary>
+        /// Writes fixup data into current ObjectDataBuilder. Caller needs to make sure ObjectDataBuilder is
+        /// at correct offset before writing.
+        /// </summary>
+        /// <returns>Bytes written</returns>
+        public abstract void WriteData(ref ObjectDataBuilder builder, NodeFactory factory);
+    }
+
+    public class PreInitTypeFixupInfo : PreInitFixupInfo
+    {
+        public TypeDesc TypeFixup { get; }
+
+        public PreInitTypeFixupInfo(int offset, TypeDesc type)
+            :base(offset)
+        {
+            TypeFixup = type;
+        }
+
+        public override void WriteData(ref ObjectDataBuilder builder, NodeFactory factory)
+        {
+            builder.EmitPointerReloc(factory.NecessaryTypeSymbol(TypeFixup));
+        }
+    }
+
+    public class PreInitMethodFixupInfo : PreInitFixupInfo
+    {
+        public MethodDesc MethodFixup { get; }
+
+        public PreInitMethodFixupInfo(int offset, MethodDesc method)
+            : base(offset)
+        {
+            if (method.HasInstantiation)
```

Comment:
My previous comment was about methods with genericness in general. This will still assert the compiler if e.g. this is a static method on a generic type. (You can call this a nit since this is not a user exposed feature, but if the compiler asserts, it will likely be someone like me called in to investigate.)